### ISSUE TITLE
Suggestions on methods for printing messages

### ISSUE THREAD

**Issue by Feng1231** [nus-cs2103-AY2425S1/forum#232]
Hi! I'm trying to improve my code. I am looking through the UI's printing message methods.

There are two options for printing:

1st:
Having one method for each print case, eg.
```java
printEventCommandError()
printDeleteCommandError()
...
```

2nd:
Having one unified printing method that takes in the error message as parameter
`printMessage(String message)
`
Would like to ask for some opinions on which option seems better. Any input is appreciated 🙂

**Comment by JoanneHing** [comment]
For me I don't actually has method that print out different command error, I had a parent class for Exception (`ToothlessException`) and some inherited class that represent different type of error such as `MissingIndexException`, `MissingDescriptionException` and etc. Each has used the `super(String errorMessage)` in their own constructor to pass the error message to be printed. And I used `try` `catch` in my `main()` method to print the message from the exception via `e.getMessage()`. Not sure is this the best way to do it but it reduce the need to create lots of methods in the `Ui` class to print the messages. Hope it will helps!

**Comment by Feng1231** [comment]
> For me I don't actually has method that print out different command error, I had a parent class for Exception (`ToothlessException`) and some inherited class that represent different type of error such as `MissingIndexException`, `MissingDescriptionException` and etc. Each has used the `super(String errorMessage)` in their own constructor to pass the error message to be printed. And I used `try` `catch` in my `main()` method to print the message from the exception via `e.getMessage()`. Not sure is this the best way to do it but it reduce the need to create lots of methods in the `Ui` class to print the messages. Hope it will helps!

May I clarify what you did under the main() method was this?

```java
catch (ToothlessException e) {
    System.out.println(e.getMessage());
}
```

**Comment by RezwanAhmed123** [comment]
## CLI-App

I had a dedicated printing class at first when I was still working on the CLI-App. The excerpt below is of my printer class from then:

``` java
public class FormattedPrinting {

    public static void formatPrint(String message) {
        String indentation = "   ";
        String separatorLine = indentation + "______________________________________________________";
        StringBuilder outputMessage = new StringBuilder(separatorLine);
        String[] messageLines = message.split("\n");
        for (String line : messageLines) {
            outputMessage.append("\n")
                    .append(indentation)
                    .append(line);
        }
        outputMessage.append("\n")
                .append(separatorLine);
        System.out.println(outputMessage.toString());
    }

    public static void welcome() {
        String welcomeMessage = "Hello I'm Mortal Reminder!\n"
                + "What can I do for you?";
        formatPrint(welcomeMessage);

    }
}
```


The first method allowed all of my print classes to have a consistent format in the printing styles. This class also mostly consisted of static methods (as seen in the second method) which could be called from other classes without instantiating an object of said class.

## GUI-App
Later on, specifically after I moved to GUI-App, I realised this method was kind of becoming redundant as the GUI basically handled all of the formatting for me. As such, I incorporated an error class which then helped to pass in any error messages. An excerpt is shown below:

``` java
/**
     * Retrieves the task at the specified index from the list.
     * <p>
     * If the index is out of bounds or the list is empty, an appropriate message is printed
     * and {@code null} is returned.
     *
     * @param index the index of the task to retrieve.
     * @return the {@link Task} at the specified index, or {@code null} if the index is invalid.
     * @throws MortalReminderException if the index is invalid in the list.
     */
    public Task getTask(int index) throws MortalReminderException {
        try {
            return this.taskList.get(index);
        } catch (IndexOutOfBoundsException e) {
            if (this.getSize() == 0) {
                throw new MortalReminderException("List is empty!");
            }
            throw new MortalReminderException("Invalid task number!\n"
                    + "Please input a number between 1 and "
                    + this.getSize());
        }
    }
```

In spirit of laziness and not wanting to re-code everything, I changed return type in the printer class to strings. I kept the first method of the printer class as the following:

``` java
/**
     * Prints out the formatted output of the chatbot to the user.
     * An indentation is added to each line and the separator lines are added at the top and bottom of each
     * line in the message.
     *
     * @param message output message from chatbot to be printed out to the user
     * @return a version of confirmation message with specific modifications.
     */
    public static String getResponse(String message) {
        return message;
    }
```

and to make it "useful" maybe when I am done with this weeks tasks add a list of "personality quirks" which i can add at random to the start of each output message.

## How were these errors printed?
Basically, I had the main class in my project to do a try catch block like so:

``` java
/**
     * Initialises the execution of a command and returns the feedback message from the processor.
     *
     * @param command command created using the parser.
     * @return the string response from the chatbot after command has been executed.
     */
    public String executeCommand(Command command) {
        try {
            return processor.handleCommand(command, taskList);
        } catch (MortalReminderException e) {
            return e.getMessage();
        }
    }
```

**Comment by ahmadsyuaib** [comment]
> > For me I don't actually has method that print out different command error, I had a parent class for Exception (`ToothlessException`) and some inherited class that represent different type of error such as `MissingIndexException`, `MissingDescriptionException` and etc. Each has used the `super(String errorMessage)` in their own constructor to pass the error message to be printed. And I used `try` `catch` in my `main()` method to print the message from the exception via `e.getMessage()`. Not sure is this the best way to do it but it reduce the need to create lots of methods in the `Ui` class to print the messages. Hope it will helps!
> 
> May I clarify what you did under the main() method was this?
> 
> ```java
> catch (ToothlessException e) {
>     System.out.println(e.getMessage());
> }
> ```

Hi there! Yes, that would be how I implemented the exceptions too. If you have not added the JavaFX, it would be wherever you have your `while` loop. If you have added JavaFX, you could add the `try-catch` block in the `getResponse()` method and instead of printing, you can return the error message as a String. Both this method would allow you to continue using the application even after the error is thrown.

**Comment by JoanneHing** [comment]
> > For me I don't actually has method that print out different command error, I had a parent class for Exception (`ToothlessException`) and some inherited class that represent different type of error such as `MissingIndexException`, `MissingDescriptionException` and etc. Each has used the `super(String errorMessage)` in their own constructor to pass the error message to be printed. And I used `try` `catch` in my `main()` method to print the message from the exception via `e.getMessage()`. Not sure is this the best way to do it but it reduce the need to create lots of methods in the `Ui` class to print the messages. Hope it will helps!
> 
> May I clarify what you did under the main() method was this?
> 
> ```java
> catch (ToothlessException e) {
>     System.out.println(e.getMessage());
> }
> ```

Yup correct! You may refer to my PR also if you are interested to know more about my code organization!

**Comment by Feng1231** [comment]
Thanks all for the valuable inputs! It seems like I am doing the same thing as well. So I guess I will leave it as it is, and not make UI have multiple methods, each printing a different string.
